File: Top_100_Liked_Questions/49._Group_Anagrams/solution.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # using count as hash
        d = {}
        for s in strs:
            count = [0] * 26
            for ch in s:
                count[ord(ch) - ord('a')] += 1
                
            key = tuple(count)
            if key in d:
                d[key].append(s)
            else:
                d[key] = [s]
                
        return d.values()
    
    
    def groupAnagrams3(self, strs: List[str]) -> List[List[str]]:
        # sorting method - O(n*m*log(m))
        d = {}
        for s in strs:
            sortedStr = ''.join(sorted(s))
            if sortedStr in d:
                d[sortedStr].append(s)
            else:
                d[sortedStr] = [s]
                
        return d.values()
    
    
    def groupAnagrams2(self, strs: List[str]) -> List[List[str]]:
        # brute force - O(n**2 + n*m)
        ans = []
        for s in strs:
            curr = collections.Counter(s)
            found = False
            for lst in ans:
                if lst[0] == curr:
                    lst.append(s)
                    found = True
                    break
            
            if not found:
                ans.append([curr, s])
        
        for lst in ans:
            del lst[0]
        
        return ans

File: Top_100_Liked_Questions/49._Group_Anagrams/solution.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    // hash count method
    let hash = {};
    for (let s of strs) {
        let count = new Array(26).fill(0);
        for (let i=0; i<s.length; i++) {
            count[s.charCodeAt(i) - "a".charCodeAt(0)] += 1;
        }
        if (hash.hasOwnProperty(count)) {
            hash[count].push(s);
        } else {
            hash[count] = [s];
        }
    }
    
    ans = [];
    Object.values(hash).forEach(v => ans.push(v));
    
    return ans;
};

var groupAnagrams2 = function(strs) {
    // sorting method
    let hash = {};
    for (let s of strs) {
        let key = s.split("").sort()
        if (hash.hasOwnProperty(key)) {
            hash[key].push(s);
        } else {
            hash[key] = [s];
        }
    }
    
    ans = [];
    Object.values(hash).forEach(v => ans.push(v));
    
    return ans;
};